File: 121-BestTimeToBuyAndSellStock/121-BestTimeToBuyAndSellStock.py
# Last updated: 8/13/2025, 8:22:33 PM
class Solution(object):
    def maxProfit(self, prices):
        pricesLength = len(prices)
        maxPrefixPrices = [0] * pricesLength
        maxSuffixPrices = [0] * pricesLength
        profit = 0
        maxProfit = -sys.maxint
        
        maxSuffixPrices[pricesLength-1] = prices[pricesLength-1]

        for i in range(len(prices)-2 , -1 , -1) :
            maxSuffixPrices[i] = max(maxSuffixPrices[i+1] , prices[i])
        
        print(maxSuffixPrices)

        maxPrefixPrices[0] = prices[0]

        for i in range(1,len(prices)) :
            maxPrefixPrices[i] = max(maxPrefixPrices[i-1] , prices[i])

        print(maxPrefixPrices)
        
        for i in range(pricesLength-1) :
            profit = maxSuffixPrices[i] - prices[i]
            maxProfit = max(profit , maxProfit)
        
        if maxProfit < 0 :
            return 0
        
        return maxProfit
        

File: 121-BestTimeToBuyAndSellStock/121-BestTimeToBuyAndSellStock.java
// Last updated: 8/13/2025, 8:22:38 PM
class Solution {
    public int maxProfit(int[] prices) {
        int prefixMin[] = new int[prices.length];
        int suffixMax[] = new int[prices.length];

        prefixMin[0] = prices[0];
        for(int i=1;i<prices.length;i++){
            prefixMin[i] = Math.min(prefixMin[i-1],prices[i]);
        }

        suffixMax[prices.length-1] = prices[prices.length-1];
        for(int i=prices.length-2; i >=0;i--){
            suffixMax[i] = Math.max(suffixMax[i+1],prices[i]);
        }
        int profit = 0;
        int maxProfit = 0;
        for(int i=0;i<prices.length;i++){
            if((suffixMax[i] - prefixMin[i])>0){
                profit = suffixMax[i] - prefixMin[i];
            }
            maxProfit = Math.max(profit , maxProfit);
        }
        return maxProfit;
    }
}